klyshko-operator/controllers/tuplegenerationjob_controller.go

```diff
@@ -0,0 +1,443 @@
+type TupleGenerationJobReconciler struct {
+	client.Client
+	Scheme          *runtime.Scheme
+	EtcdClient      *clientv3.Client
+	rosterWatcherCh clientv3.WatchChan
+	CastorClient    *castor.Client
+	Logger          logr.Logger
+}
+
+// NewTupleGenerationJobReconciler creates a TupleGenerationJobReconciler.
+func NewTupleGenerationJobReconciler(client client.Client, scheme *runtime.Scheme, etcdClient *clientv3.Client, castorClient *castor.Client) *TupleGenerationJobReconciler {
+	r := &TupleGenerationJobReconciler{
+		Client:          client,
+		Scheme:          scheme,
+		EtcdClient:      etcdClient,
+		rosterWatcherCh: etcdClient.Watch(context.Background(), rosterKey, clientv3.WithPrefix()), // TODO Close channel?
+		CastorClient:    castorClient,
+		Logger:          logr.DiscardLogger{},
+	}
+	go r.handleWatchEvents()
+	return r
+}
+
+//+kubebuilder:rbac:groups=klyshko.carbnyestack.io,resources=tuplegenerationjobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=klyshko.carbnyestack.io,resources=tuplegenerationjobs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=klyshko.carbnyestack.io,resources=tuplegenerationjobs/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+// Reconcile compares the actual state of TupleGenerationJob resources to their desired state and performs actions to
+// bring the actual state closer to the desired one.
+func (r *TupleGenerationJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	// Assigning logger to field to allow access from etcd event handling logic below
+	r.Logger = log.FromContext(ctx)
+
+	jobKey := RosterKey{
+		req.NamespacedName,
+	}
+	logger := r.Logger.WithValues("Job.Key", jobKey)
+	logger.V(logging.DEBUG).Info("Reconciling tuple generation jobs")
+
+	// Cleanup if job has been deleted
+	job := &klyshkov1alpha1.TupleGenerationJob{}
+	err := r.Get(ctx, req.NamespacedName, job)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Job resource not available -> has been deleted
+			_, err := r.EtcdClient.Delete(ctx, jobKey.ToEtcdKey())
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to delete roster for job %v: %w", req.Name, err)
+			}
+			logger.V(logging.DEBUG).Info("Roster deleted")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, fmt.Errorf("failed to read resource for job %v: %w", req.Name, err)
+	}
+	logger.V(logging.TRACE).Info("Job exists already")
+
+	// Create roster if not existing (no etcd transaction needed as remote job creation is triggered by roster creation)
+	resp, err := r.EtcdClient.Get(ctx, jobKey.ToEtcdKey())
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to read resource for roster with key %v for task %v: %w", jobKey, req.Name, err)
+	}
+	if resp.Count == 0 {
+		playerID, err := localPlayerID(ctx, &r.Client, req.Namespace)
```

Comment:
Two comments here:

1. The K8s API server should be the single source of truth. The K8s API client anyway has a caching mechanism built in, so that reading of the config map should be reasonably cheap.
2. The way it is implemented here is legacy from day 1. The upcoming CS operator will provide CRDs for VCPs and VC containing the information that is currently stored in the config map. So I would suggest not to spend the effort to optimize anything unless it really hurts so much that it can't wait for the operator to arrive.

Fine for you?